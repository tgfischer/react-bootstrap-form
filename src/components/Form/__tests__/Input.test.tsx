import React from "react";
import * as yup from "yup";
import { render, fireEvent, waitFor } from "@testing-library/react";

import { Form } from "../../../index";
import { SampleForm } from "./SampleForm";

describe("Input tests", () => {
  const handleSubmit = jest.fn();
  const handleChange = jest.fn();
  const message = "Hello, World!";

  it("should enter text in the input, and submit the form", async () => {
    const { getByLabelText, getByText } = render(
      <SampleForm initialValues={{ input: "" }} onSubmit={handleSubmit}>
        <Form.Input name="input" label="Input field" />
      </SampleForm>
    );

    fireEvent.change(getByLabelText("Input field"), {
      target: { value: message }
    });
    fireEvent.click(getByText("Submit"));

    await waitFor(() =>
      expect(handleSubmit).toHaveBeenCalledWith(
        { input: message },
        expect.any(Object)
      )
    );
  });

  it("should show the error feedback message", async () => {
    const { container, getByLabelText, getByText } = render(
      <SampleForm
        initialValues={{ input: "" }}
        validationSchema={yup
          .object({
            input: yup.string().max(1).required()
          })
          .required()}
        onSubmit={handleSubmit}
      >
        <Form.Input name="input" label="Input field" />
      </SampleForm>
    );

    fireEvent.change(getByLabelText("Input field"), {
      target: { value: "batman" }
    });
    fireEvent.click(getByText("Submit"));

    await waitFor(() =>
      expect(container.querySelector(".invalid-feedback")).toHaveTextContent(
        "input must be at most 1 characters"
      )
    );
    expect(handleSubmit).not.toHaveBeenCalled();
  });

  it("should not show the error feedback if the input has not been touched", async () => {
    const { container, getByLabelText, getByText } = render(
      <SampleForm
        initialValues={{ input1: "", input2: "" }}
        validationSchema={yup
          .object({
            input1: yup.string().required(),
            input2: yup.string().required()
          })
          .required()}
        onSubmit={handleSubmit}
      >
        <Form.Input name="input1" label="Input 1" />
        <Form.Input name="input2" label="Input 2" />
      </SampleForm>
    );

    // Ensure the error feedback is empty
    fireEvent.change(getByLabelText("Input 2"), {
      target: { value: "batman" }
    });
    expect(container.firstChild).toMatchSnapshot();

    // Ensure that the error feedback is set
    fireEvent.click(getByText("Submit"));
    await waitFor(() =>
      expect(container.querySelector(".invalid-feedback")).toHaveTextContent(
        "input1 is a required field"
      )
    );
  });

  it("should call on change event", async () => {
    const { getByLabelText } = render(
      <SampleForm initialValues={{ input: "" }} onSubmit={handleSubmit}>
        <Form.Input name="input" label="Input field" onChange={handleChange} />
      </SampleForm>
    );

    fireEvent.change(getByLabelText("Input field"), {
      target: { value: "batman" }
    });

    await waitFor(() =>
      expect(handleChange).toHaveBeenCalledWith(expect.any(Object))
    );
  });
});
