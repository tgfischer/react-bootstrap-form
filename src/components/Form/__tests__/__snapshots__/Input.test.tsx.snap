// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input tests should not show the error feedback if the input has not been touched 1`] = `
<form
  class=""
>
  <div
    class="form-group"
    name="input1"
  >
    <label
      class="form-label"
      for="input1"
    >
      Input 1
    </label>
    <input
      class="form-control"
      id="input1"
      name="input1"
      value=""
    />
    <div
      class="invalid-feedback"
    />
  </div>
  <div
    class="form-group"
    name="input2"
  >
    <label
      class="form-label"
      for="input2"
    >
      Input 2
    </label>
    <input
      class="form-control"
      id="input2"
      name="input2"
      value="batman"
    />
    <div
      class="invalid-feedback"
    />
  </div>
  <div
    class="d-flex justify-content-end"
  >
    <button
      class="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
